64_FFT/main.cpp

```diff
@@ -0,0 +1,307 @@
+{
+	using device_base_t = application_templates::MonoDeviceApplication;
+	using asset_base_t = application_templates::MonoAssetManagerAndBuiltinResourceApplication;
+
+	smart_refctd_ptr<IGPUComputePipeline> m_pipeline;
+
+	smart_refctd_ptr<nbl::video::IUtilities> m_utils;
+
+	nbl::video::StreamingTransientDataBufferMT<>* m_upStreamingBuffer;
+	StreamingTransientDataBufferMT<>* m_downStreamingBuffer;
+
+	// These are Buffer Device Addresses
+	uint64_t m_upStreamingBufferAddress;
+	uint64_t m_downStreamingBufferAddress;
+
+	// You can ask the `nbl::core::GeneralpurposeAddressAllocator` used internally by the Streaming Buffers give out offsets aligned to a certain multiple (not only Power of Two!)
+	uint32_t m_alignment;
+
+	// The pool cache is just a formalized way of round-robining command pools and resetting + reusing them after their most recent submit signals finished.
+	// Its a little more ergonomic to use if you don't have a 1:1 mapping between frames and pools.
+	smart_refctd_ptr<nbl::video::ICommandPoolCache> m_poolCache;
+
+	// This example really lets the advantages of a timeline semaphore shine through!
+	smart_refctd_ptr<ISemaphore> m_timeline;
+	uint64_t m_iteration = 0;
+	constexpr static inline uint64_t MaxIterations = 1;
+
+public:
+	// Yay thanks to multiple inheritance we cannot forward ctors anymore
+	StreamingAndBufferDeviceAddressApp(const path& _localInputCWD, const path& _localOutputCWD, const path& _sharedInputCWD, const path& _sharedOutputCWD) :
+		system::IApplicationFramework(_localInputCWD, _localOutputCWD, _sharedInputCWD, _sharedOutputCWD) {}
+
+	// we stuff all our work here because its a "single shot" app
+	bool onAppInitialized(smart_refctd_ptr<ISystem>&& system) override
+	{
+		// Remember to call the base class initialization!
+		if (!device_base_t::onAppInitialized(smart_refctd_ptr(system)))
+			return false;
+		if (!asset_base_t::onAppInitialized(std::move(system)))
+			return false;
+
+		// this time we load a shader directly from a file
+		smart_refctd_ptr<IGPUShader> shader;
+		{
+			IAssetLoader::SAssetLoadParams lp = {};
+			lp.logger = m_logger.get();
+			lp.workingDirectory = ""; // virtual root
+			auto assetBundle = m_assetMgr->getAsset("app_resources/shader.comp.hlsl", lp);
+			const auto assets = assetBundle.getContents();
+			if (assets.empty())
+				return logFail("Could not load shader!");
+
+			// lets go straight from ICPUSpecializedShader to IGPUSpecializedShader
+			auto source = IAsset::castDown<ICPUShader>(assets[0]);
+			// The down-cast should not fail!
+			assert(source);
+
+			// this time we skip the use of the asset converter since the ICPUShader->IGPUShader path is quick and simple
+			shader = m_device->createShader(source.get());
+			if (!shader)
+				return logFail("Creation of a GPU Shader to from CPU Shader source failed!");
+		}
```

Comment:
I'd dynamically make a shader source with following contents

```cpp
#define _NBL_HLSL_WORKGROUP_SIZE_ ...whatever when you produce the runtime string...

#include "app_resources/shader.comp.hlsl"
```



at alternative is `createOVerriden` but its more C++ code to put together